### PR TITLE
Drivetrain

### PR DESCRIPTION
**Pull request by jaredgoodman03**

### CODE REVIEW

File: 2021-offseason/src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -12,4 +14,21 @@
  * <p>It is advised to statically import this class (or one of its inner classes) wherever the
  * constants are needed, to reduce verbosity.
  */
-public final class Constants {}
+public final class Constants {
+    public static final class kDrivetrain{
+        public static final int frontLeft_ID = 18;
+        public static final int frontRight_ID = 15;
+        public static final int backLeft_ID = 1;
+        public static final int backRight_ID = 14;
```
Comment (jaredgoodman03) [inline 2021-offseason/src/main/java/frc/robot/Constants.java:22]: For all final values in java you should name your variables upper case with underscores in between each word
`public static final int FRONT_LEFT_ID`

File: 2021-offseason/src/main/java/frc/robot/Constants.java [hunk 2]
```diff
@@ -12,4 +14,21 @@
  * <p>It is advised to statically import this class (or one of its inner classes) wherever the
  * constants are needed, to reduce verbosity.
  */
-public final class Constants {}
+public final class Constants {
+    public static final class kDrivetrain{
```
Comment (jaredgoodman03) [inline 2021-offseason/src/main/java/frc/robot/Constants.java:18]: Add a space in between `kDrivetrian` and `{`

File: 2021-offseason/src/main/java/frc/robot/Constants.java [hunk 3]
```diff
@@ -12,4 +14,21 @@
...
+        public static final int backLeft_ID = 1;
+        public static final int backRight_ID = 14;
+
+        public static final CANSparkMaxLowLevel.MotorType MOTOR_TYPE = CANSparkMaxLowLevel.MotorType.kBrushless;
+
+        public static final int CURRENT_LIMIT = 35;
+        public static final int OPEN_LOOP_RAMP = 0;
+        public static final double SLOW_SPEED = 0.1;
+    }
+    public static final class kOI{//koi fish
```
Comment (jaredgoodman03) [inline 2021-offseason/src/main/java/frc/robot/Constants.java:30]: space between `kOI` and `{`

File: 2021-offseason/src/main/java/frc/robot/OI.java [hunk 4]
```diff
@@ -1,5 +1,15 @@
...
 import edu.wpi.first.wpilibj.XboxController;
-import edu.wpi.first.wpilibj.controller.*;
-public class OI {
+import edu.wpi.first.wpilibj.GenericHID.Hand;
 
-}
\ No newline at end of file
+
+public class OI {
+    public static double getTriggers(XboxController controller){
```
Comment (jaredgoodman03) [inline 2021-offseason/src/main/java/frc/robot/OI.java:8]: spaces in this file too

File: 2021-offseason/src/main/java/frc/robot/RobotContainer.java [hunk 5]
```diff
@@ -22,9 +24,19 @@
...
+  private static final XboxController driveController = new XboxController(Constants.kOI.DRIVE_CONTROLLER);
+  
+
   /** The container for the robot. Contains subsystems, OI devices, and commands. */
   public RobotContainer() {
+    s_Drivetrain = new Drivetrain();
+    
     // Configure the button bindings
+    s_Drivetrain.setDefaultCommand(new RunCommand(
+      () -> s_Drivetrain.curveDrive(OI.getTriggers(driveController), OI.getLeftStick(driveController), driveController.getXButton()), s_Drivetrain
```
Comment (jaredgoodman03) [inline 2021-offseason/src/main/java/frc/robot/RobotContainer.java:38]: split onto more lines

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    backLeft.setOpenLoopRampRate(Constants.kDrivetrain.OPEN_LOOP_RAMP);
+    backRight.setOpenLoopRampRate(Constants.kDrivetrain.OPEN_LOOP_RAMP);
+
+    frontLeft.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+    frontRight.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+    backLeft.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+    backRight.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+  }
+
+  public void curveDrive(double linearVelocity, double angularVelocity, boolean isQuickturn){
```
Comment (jaredgoodman03) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:47]: spacing here

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+import com.revrobotics.CANSparkMax;
+
+import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class Drivetrain extends SubsystemBase {
+  private CANSparkMax frontLeft, frontRight, backLeft, backRight;
+
```
Comment (jaredgoodman03) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:15]: remove this empty line it's ugly

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  private DifferentialDrive difDrive;
+  private boolean driveInverted;
+  /** Creates a new ExampleSubsystem. */
+  public Drivetrain() {
+    driveInverted = false;
```
Comment (jaredgoodman03) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:20]: put this in constants

File: 2021-offseason/src/main/java/frc/robot/Constants.java [hunk 9]
```diff
@@ -12,4 +14,21 @@
  * <p>It is advised to statically import this class (or one of its inner classes) wherever the
  * constants are needed, to reduce verbosity.
  */
-public final class Constants {}
+public final class Constants {
+    public static final class kDrivetrain{
+        public static final int frontLeft_ID = 18;
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/Constants.java:19]: nit: style: names should be capitalized to follow constant naming, e.g. `FRONT_LEFT_ID`

File: 2021-offseason/src/main/java/frc/robot/Constants.java (hunk 3 above)
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/Constants.java:30]: nit: style: space before the opening parenthesis

Comment (KTOmega) [reply 2021-offseason/src/main/java/frc/robot/Constants.java:30]: also please put the comment on a new line

File: 2021-offseason/src/main/java/frc/robot/OI.java (hunk 4 above)
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/OI.java:8]: nit: style: space before opening parenthesis (also line 12)

File: 2021-offseason/src/main/java/frc/robot/RobotContainer.java [hunk 10]
```diff
@@ -22,9 +24,19 @@
 
   private final ExampleCommand m_autoCommand = new ExampleCommand(m_exampleSubsystem);
 
+  private final Drivetrain s_Drivetrain;
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/RobotContainer.java:27]: nit: naming: variable name should be more snake-casey, i.e. `s_drivetrain`

File: 2021-offseason/src/main/java/frc/robot/RobotContainer.java [hunk 11]
```diff
@@ -22,9 +24,19 @@
...
+  
+
   /** The container for the robot. Contains subsystems, OI devices, and commands. */
   public RobotContainer() {
+    s_Drivetrain = new Drivetrain();
+    
     // Configure the button bindings
+    s_Drivetrain.setDefaultCommand(new RunCommand(
+      () -> s_Drivetrain.curveDrive(OI.getTriggers(driveController), OI.getLeftStick(driveController), driveController.getXButton()), s_Drivetrain
+      ));
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/RobotContainer.java:39]: nit: style: spacing
```suggestion
    s_Drivetrain.setDefaultCommand(new RunCommand(
      () -> s_Drivetrain.curveDrive(OI.getTriggers(driveController), OI.getLeftStick(driveController), driveController.getXButton()),
      s_Drivetrain));
```

File: 2021-offseason/src/main/java/frc/robot/RobotContainer.java [hunk 12]
```diff
@@ -43,6 +55,6 @@ private void configureButtonBindings() {}
    */
   public Command getAutonomousCommand() {
     // An ExampleCommand will run in autonomous
-    return m_autoCommand;
+    return m_autoCommand;
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/RobotContainer.java:58]: nit: spacing: remove the extra space

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java (hunk 8 above)
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:20]: this should be a constant in the `Constants` file

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class Drivetrain extends SubsystemBase {
+  private CANSparkMax frontLeft, frontRight, backLeft, backRight;
+
+  private DifferentialDrive difDrive;
+  private boolean driveInverted;
+  /** Creates a new ExampleSubsystem. */
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:18]: please remove this comment or replace it with a more relevant comment

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class Drivetrain extends SubsystemBase {
+  private CANSparkMax frontLeft, frontRight, backLeft, backRight;
+
+  private DifferentialDrive difDrive;
+  private boolean driveInverted;
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:17]: nit: can we try using `final` on all member variables?
```suggestion
  private final CANSparkMax frontLeft, frontRight, backLeft, backRight;

  private final DifferentialDrive difDrive;
  private final boolean driveInverted;
```

File: 2021-offseason/src/main/java/frc/robot/RobotContainer.java [hunk 15]
```diff
@@ -22,9 +24,19 @@
 
   private final ExampleCommand m_autoCommand = new ExampleCommand(m_exampleSubsystem);
 
+  private final Drivetrain s_Drivetrain;
+
+  private static final XboxController driveController = new XboxController(Constants.kOI.DRIVE_CONTROLLER);
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/RobotContainer.java:29]: why is this static?

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java (hunk 6 above)
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:47]: nit: style: space before opening parenthesis

**Review by KTOmega — CHANGES_REQUESTED** [review]
looks good, but had a couple of comments

File: 2021-offseason/src/main/java/frc/robot/Constants.java (hunk 3 above)
Comment (uniqueusername) [reply 2021-offseason/src/main/java/frc/robot/Constants.java:30]: for stuff like open loop ramp having a comment w units would be nice

File: 2021-offseason/src/main/java/frc/robot/OI.java (hunk 4 above)
Comment (uniqueusername) [reply 2021-offseason/src/main/java/frc/robot/OI.java:8]: add a method comment imo, `getTriggers` doesn't obviously subtract the left hand from right hand inputs

File: 2021-offseason/src/main/java/frc/robot/OI.java [hunk 16]
```diff
@@ -1,5 +1,15 @@
...
+        double val = controller.getTriggerAxis(Hand.kRight) - controller.getTriggerAxis(Hand.kLeft);
+        return Math.pow(val, 3);
+    }
+    public static double getLeftStick(XboxController controller){
```
Comment (uniqueusername) [inline 2021-offseason/src/main/java/frc/robot/OI.java:12]: line break between 11 and 12
```suggestion
    }
    
    public static double getLeftStick(XboxController controller){
```

File: 2021-offseason/src/main/java/frc/robot/OI.java (hunk 4 above)
Comment (jaredgoodman03) [reply 2021-offseason/src/main/java/frc/robot/OI.java:8]: Or you could additionally name this function better (e.g. getTriggerTotal)

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+import com.revrobotics.CANSparkMax;
+import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class Drivetrain extends SubsystemBase {
+  private final CANSparkMax frontLeft, frontRight, backLeft, backRight;
+  private final DifferentialDrive difDrive;
+  private final boolean driveInverted;
```
Comment (KTOmega) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:15]: does this need to be a member variable? it's only used in the constructor
furthermore, do we need to make this its own variable? you could just do `frontLeft.setInverted(!Constants.kDrivetrain.DRIVE_INVERTED);` and I don't think that would change anything or decrease code readability

File: 2021-offseason/src/main/java/frc/robot/OI.java [hunk 18]
```diff
@@ -1,15 +1,16 @@
...
-        return Math.pow(difference, 3);
+    //returns difference between Right trigger value and Left trigger value;
+    public static double getTriggers (XboxController controller) {
+        double val = controller.getTriggerAxis(Hand.kRight) - controller.getTriggerAxis(Hand.kLeft);
+        return Math.pow(val, 3);
     }
 
-    public static double getLeftJoystick(XboxController controller) {
-        return Math.pow(controller.getX(GenericHID.Hand.kLeft), 3);
+    public static double getLeftStick (XboxController controller) {
```
[the remark this answers is not in the document]
Comment (KTOmega) [reply 2021-offseason/src/main/java/frc/robot/OI.java:13]: same for line 8

**Review by KTOmega — CHANGES_REQUESTED** [review]
a few more comments...

File: 2021-offseason/src/main/java/frc/robot/OI.java [hunk 19]
```diff
@@ -1,15 +1,16 @@
...
-import edu.wpi.first.wpilibj.GenericHID;
+import edu.wpi.first.wpilibj.GenericHID.Hand;
 
 public class OI {
-    public static double getTriggerOutput(XboxController controller) {
-        double leftTrigger = controller.getTriggerAxis(GenericHID.Hand.kLeft);
-        double rightTrigger = controller.getTriggerAxis(GenericHID.Hand.kRight);
-        double difference = rightTrigger - leftTrigger;
-        return Math.pow(difference, 3);
+    //returns difference between Right trigger value and Left trigger value;
```
Comment (uniqueusername) [inline 2021-offseason/src/main/java/frc/robot/OI.java:7]: nit: space before beginning of comment
```
// returns difference between Right trigger value and Left trigger value;
```

File: 2021-offseason/src/main/java/frc/robot/RobotContainer.java [hunk 20]
```diff
@@ -60,12 +62,12 @@ private void configureButtonBindings() {
       .whenPressed(new RunCommand(s_intake::stopIntake, s_intake)); 
 
     new JoystickButton(operatorController, XboxController.Button.kX.value)
-      .whenPressed(new RunCommand(m_hopper::moveForward, m_hopper))
-      .whenReleased(new RunCommand(m_hopper::stop, m_hopper));
+      .whenPressed(new RunCommand(s_hopper::moveForward, s_hopper))
```
Comment (uniqueusername) [inline 2021-offseason/src/main/java/frc/robot/RobotContainer.java:65]: is this supposed to be in this branch lol

File: 2021-offseason/src/main/java/frc/robot/RobotContainer.java [hunk 21]
```diff
@@ -75,6 +77,6 @@ private void configureButtonBindings() {
    */
   public Command getAutonomousCommand() {
     // An ExampleCommand will run in autonomous
-    return m_autoCommand;
+    return null;
```
Comment (uniqueusername) [inline 2021-offseason/src/main/java/frc/robot/RobotContainer.java:80]: why

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+package frc.robot.subsystems;
+
+import com.revrobotics.CANSparkMax;
+import edu.wpi.first.wpilibj.drive.DifferentialDrive;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class Drivetrain extends SubsystemBase {
+  private final CANSparkMax frontLeft, frontRight, backLeft, backRight;
+  private final DifferentialDrive difDrive;
```
Comment (uniqueusername) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:14]: two fs ``diffDrive``

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java (hunk 17 above)
Comment (uniqueusername) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:15]: i'm pretty sure this already is a constant in this pr ?

File: 2021-offseason/src/main/java/frc/robot/RobotContainer.java (hunk 21 above)
Comment (KTOmega) [reply 2021-offseason/src/main/java/frc/robot/RobotContainer.java:80]: I assumed this was an artifact of the merge commit that @jaredgoodman03 made. While it is innocuous (as `m_autoCommand` was removed), we should avoid making changes like these in the future, as this change will be marked on the blame as coming from the drivetrain commit + it makes the history confusing in general. These cleanup changes should be made in a separate PR.

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java (hunk 17 above)
Comment (uniqueusername) [reply 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:15]: but also we usually have a method within the drive subsystem to change inversion. which is why we've had it as a member variable in the past

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+    backRight.setOpenLoopRampRate(Constants.kDrivetrain.OPEN_LOOP_RAMP);
+
+    frontLeft.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+    frontRight.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+    backLeft.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+    backRight.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+  }
+
+  public void curveDrive (double linearVelocity, double angularVelocity, boolean isQuickturn) {
+    difDrive.curvatureDrive(linearVelocity, angularVelocity, isQuickturn);
```
Comment (uniqueusername) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:47]: diffDrive

File: 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+    backLeft.setOpenLoopRampRate(Constants.kDrivetrain.OPEN_LOOP_RAMP);
+    backRight.setOpenLoopRampRate(Constants.kDrivetrain.OPEN_LOOP_RAMP);
+
+    frontLeft.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+    frontRight.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+    backLeft.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+    backRight.setSmartCurrentLimit(Constants.kDrivetrain.CURRENT_LIMIT);
+  }
+
+  public void curveDrive (double linearVelocity, double angularVelocity, boolean isQuickturn) {
```
Comment (uniqueusername) [inline 2021-offseason/src/main/java/frc/robot/subsystems/Drivetrain.java:46]: nit: no space between signature and parameters

**Review by uniqueusername — CHANGES_REQUESTED** [review]

**Review by KTOmega — APPROVED** [review]
lgtm

**Pull request merged**